File: leetcode/6_ZigZagConversion.java
class Solution {
    public String convert(String s, int numRows) {
        
        if(numRows==1){
            return s;
        }
        
        int space = (numRows-1)*2;
        int length = s.length();
        String ret = "";
       
        for(int i=0; i<length; i=i+space) {
            ret = ret + s.charAt(i);
        }
        
        int space2 = space;
        int space3 = 0;
        
        for(int i=1; i<numRows-1; i++) {
            space2 = space2 - 2;
            space3 = space3 + 2;
            boolean toogle = true;
            
            for(int j=i; j<length;) {
                
                ret = ret + s.charAt(j);
                
                if(toogle) {
                    j = j+space2;    
                } else {
                    j = j+space3;
                }
                toogle = !toogle;
            }    
        }
        
        for(int i=numRows-1; i<length; i=i+space) {
            ret = ret + s.charAt(i);
        }
        
        return ret;   
    }
}

File: leetcode/6_ZigZagConversion.kt
class Solution {
    fun convert(s: String, numRows: Int): String? {
        if (numRows == 1) {
            return s
        }
        val space = (numRows - 1) * 2
        var ret = ""

        for (i in s.indices step space) {
            ret = ret + s[i]
        }

        var space2 = space
        var space3 = 0

        for (i in 1 until numRows - 1) {
            space2 = space2 - 2
            space3 = space3 + 2

            var toogle = true
            var j = i

            while (j < s.length) {
                ret = ret + s[j]
                j = if (toogle) {
                    j + space2
                } else {
                    j + space3
                }
                toogle = !toogle
            }
        }

        var i = numRows - 1
        while (i < s.length) {
            ret = ret + s[i]
            i = i + space
        }
        return ret
    }
}